lib/SPIRV/OCLToSPIRV.cpp

```diff
@@ -1504,14 +1504,55 @@ void OCLToSPIRVBase::visitCallEnqueueKernel(CallInst *CI,
   // Param: Pointer to block literal
   Value *BlockLiteral = CI->getArgOperand(BlockFIdx + 1);
   Args.push_back(BlockLiteral);
+  BlockLiteral = BlockLiteral->stripPointerCasts();
+  if (auto *GV = dyn_cast<GlobalVariable>(BlockLiteral)) {
+    if (GV->hasInitializer()) {
+      Constant *Init = GV->getInitializer();
+      if (auto *Struct = dyn_cast<ConstantStruct>(Init)) {
+        Constant *SizeConst = Struct->getOperand(0);
+        Constant *AlignConst = Struct->getOperand(1);
+        auto *SizeVal = dyn_cast<ConstantInt>(SizeConst);
+        auto *AlignVal = dyn_cast<ConstantInt>(AlignConst);
+        Args.push_back(SizeVal);
+        Args.push_back(AlignVal);
+      }
+    }
+  } else {
 
-  // Param Size: Size of block literal structure
-  // Param Aligment: Aligment of block literal structure
-  // TODO: these numbers should be obtained from block literal structure
-  Type *ParamType = getBlockStructType(BlockLiteral);
-  Args.push_back(getInt32(M, DL.getTypeStoreSize(ParamType)));
-  Args.push_back(getInt32(M, DL.getPrefTypeAlign(ParamType).value()));
+    Value *Base = getUnderlyingObject(BlockLiteral);
+    Value *ParamSizeVal = nullptr;
+    Value *ParamAlignVal = nullptr;
+
+    for (User *U : Base->users()) {
+      if (auto *GEP = dyn_cast<GetElementPtrInst>(U)) {
+        if (GEP->getNumIndices() < 2)
+          continue;
+        auto *CI1 = dyn_cast<ConstantInt>(GEP->getOperand(2));
+        if (!CI1)
+          continue;
+        uint64_t FieldIndex = CI1->getZExtValue();
+        for (User *GEPUser : GEP->users()) {
+          if (auto *Store = dyn_cast<StoreInst>(GEPUser)) {
+            if (FieldIndex == 0 && !ParamSizeVal)
```

Comment:
This looks fishy, why do we need to ensure, that `ParamSizeVal` was not previously assigned? Is it a possible situation, when we store to the same GEP twice for block literals?